### PR TITLE
master_MDL-71585-58 Refactoring manage categories

### PR DESCRIPTION
**Pull request by guillogo**
PR:

1. Refactor manage categories.
2. Move unit test.
3. Add plugin navigation.

### CODE REVIEW

File: question/bank/managecategories/classes/question_category_object.php [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+
+        $this->initialize($page, $contexts, $currentcat, $defaultcategory, $todelete, $addcontexts);
+    }
+
+    /**
+     * Old syntax of class constructor. Deprecated in PHP7.
+     *
+     * @deprecated since Moodle 3.1
+     */
+    public function question_category_object($page, $pageurl, $contexts, $currentcat, $defaultcategory, $todelete, $addcontexts) {
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_object.php:124]: do we need this deprecated one here? its already there in the old one and any legacy code is not pointing to this code.

File: question/bank/managecategories/classes/question_category_object.php [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    /**
+     * Displays the user interface.
+     *
+     */
+    public function display_user_interface() : void {
+
+        // Interface for editing existing categories.
+        $this->output_edit_lists();
+
+        echo '<br />';
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_object.php:171]: please use html_writer instead

File: question/bank/managecategories/classes/question_category_object.php [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+        // Interface for adding a new category.
+        $this->output_new_table();
+        echo '<br />';
```
Comment (safatshahin) [inline question/bank/managecategories/classes/question_category_object.php:174]: please use html_writer instead

File: question/bank/managecategories/classes/form/question_category_edit_form.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+        $contexts = $this->_customdata['contexts'];
+        $currentcat = $this->_customdata['currentcat'];
+
+        $mform->addElement('header', 'categoryheader', get_string('addcategory', 'question'));
+
+        $mform->addElement('questioncategory', 'parent', get_string('parentcategory', 'question'),
+                array('contexts' => $contexts, 'top' => true, 'currentcat' => $currentcat, 'nochildrenof' => $currentcat));
+        $mform->setType('parent', PARAM_SEQUENCE);
+        if (question_is_only_child_of_top_category_in_context($currentcat)) {
```
Comment (safatshahin) [inline question/bank/managecategories/classes/form/question_category_edit_form.php:59]: please have a look at this function, looks like its exclusive to categories

**Review by safatshahin — CHANGES_REQUESTED** [review]
Hi @guillogo 
Thank you for working in this. I have made the comments in the code.
Some points i wanted to mention:
1. Moving events from core to the plugin, there are some events exclusive to the plugin, also there are some tests in the plugin, related to those, whats your thoughts about moving them to the plugin?
2. Naming the test classes and also merging them: managecategories_test has got the tests which are related to category object, where category object has got a test file, question_category_object_test, i think those can be moved to one. questionlib_test should be named specific to this plugin as tests the helper methods, maybe like qbank_managecategories_helper_test or something related.

Thanks

File: mod/quiz/edit.php [hunk 5]
```diff
@@ -45,7 +45,8 @@
 require_once($CFG->dirroot . '/mod/quiz/locallib.php');
 require_once($CFG->dirroot . '/mod/quiz/addrandomform.php');
 require_once($CFG->dirroot . '/question/editlib.php');
-require_once($CFG->dirroot . '/question/category_class.php');
+
+use qbank_managecategories\question_category_object;
```
[the remark this answers is not in the document]
Comment (guillogo) [reply mod/quiz/edit.php:49]: Removed, thank you Safat.

File: question/bank/managecategories/classes/question_category_list.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                                $page = 0, $pageparamname = 'page', $itemsperpage = 20, $context = null) {
+        parent::__construct('ul', '', $editable, $pageurl, $page, 'cpage', $itemsperpage);
+        $this->context = $context;
+    }
+
+    /**
+     * Set the array of records of list items.
+     */
+    public function get_records() : void {
+        $this->records = get_categories_for_contexts($this->context->id, $this->sortby);
```
[the remark this answers is not in the document]
Comment (guillogo) [reply question/bank/managecategories/classes/question_category_list.php:79]: Done.

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+                            ' ' . $category->idnumber, 'badge badge-primary');
+        }
+        $text .= ' (' . $category->questioncount . ')';
+        $item .= html_writer::tag('b', html_writer::link($questionbankurl, $text,
+                        ['title' => $editqestions]) . ' ');
+        $item .= format_text($category->info, $category->infoformat,
+                array('context' => $this->parentlist->context, 'noclean' => true));
+
+        // Don't allow delete if this is the top category, or the last editable category in this context.
+        if ($category->parent && !question_is_only_child_of_top_category_in_context($category->id)) {
```
[the remark this answers is not in the document]
Comment (guillogo) [reply question/bank/managecategories/classes/question_category_list_item.php:111]: You're right. Moved:
question_is_only_child_of_top_category_in_context
question_is_top_category
question_can_delete_cat

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            $deleteurl = new moodle_url($this->parentlist->pageurl, array('delete' => $this->id, 'sesskey' => sesskey()));
+            $item .= html_writer::link($deleteurl,
+                    $OUTPUT->pix_icon('t/delete', $str->delete),
+                    array('title' => $str->delete));
+        }
+
+        return $item;
```
[the remark this answers is not in the document]
Comment (guillogo) [reply question/bank/managecategories/classes/question_category_list_item.php]: Thank you Safat, I've migrated the `html_writer::` code to a Mustache now.

File: question/bank/managecategories/classes/question_category_list.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (guillogo) [reply question/bank/managecategories/classes/question_category_list.php]: function was moved. Thank you.

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    public function item_html($extraargs = array()) : string {
+        global $CFG, $OUTPUT;
+        $str = $extraargs['str'];
+        $category = $this->item;
+
+        $editqestions = get_string('editquestions', 'question');
+
+        // Each section adds html to be displayed as part of this list item.
+        $questionbankurl = new moodle_url('/question/edit.php', $this->parentlist->pageurl->params());
```
[the remark this answers is not in the document]
Comment (guillogo) [reply question/bank/managecategories/classes/question_category_list_item.php]: Hi  Safat,

This is the list of functions that weren't moved as they are called by different places.


question_get_default_category
question_get_top_category
question_get_top_categories_for_contexts
question_make_default_categories
question_categorylist

Thank you.

**Comment by guillogo** [comment]
Hi @safatshahin 

Thank you for reviewing, can you have a look again please. Done:

1. Refactor editlib.php functions.
2. Unit test merge.
3. Replace list item display with a template.
4. Refactor part of category functions in questionlib.php.
5. Get parent node via navigation lib.

These functions were'nt migrated as are called by different core parts. Can you please take a look if it is worth moving them by changing different parts in core:

- question_get_default_category
- question_get_top_category
- question_get_top_categories_for_contexts
- question_make_default_categories
- question_categorylist

Thank you,
Guillermo

**Comment by safatshahin** [comment]
Hi @guillogo 
All good for now, there are some functions which seems that they're used in different places as you said, which seems like some fundamental functions not directly one, but lets keep it there for now.

**Pull request merged**